File: ml-benchmark/fusion/identities/page5.f
Config SampleSize = 5;
Config NonLinear = true;

Inductive List = nil Unit | cons {Int, List};
Inductive NList = single List | ncons {List, NList};

head = \xs: NList.
  match xs with
    single w -> w
  | ncons {h, t} -> h
  end;

map = \f: List -> Int. fix (
  \g: NList -> List. \xs: NList.
  match xs with
    single w -> cons {f w, nil unit}
  | ncons {h, t} -> cons {f h, g t}
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

product = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 1
  | cons {h, t} -> * h (f t)
  end
);

tails = fix (
  \f: List -> Compress NList. \xs: List.
  match xs with
    nil _ -> single xs
  | cons {h, t} -> ncons {xs, f t}
  end
);

main = \xs: List. sum (map product (tails xs));

File: ml-benchmark/fusion/identities/page5.ml
type list_ = Nil of unit | Cons of int * list_
type nlist = Single of list_ | Ncons of list_ * nlist

let head xs =
  match xs with
  | Single w -> w
  | Ncons (h, t) -> h

let map f =
  let rec g xs =
    match xs with
    | Single w -> Cons (f w, Nil ())
    | Ncons (h, t) -> Cons (f h, g t)
  in
  g

let rec sum xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> h + sum t

let rec product xs =
  match xs with
  | Nil _ -> 1
  | Cons (h, t) -> h * product t

let rec tails xs =
  match xs with
  | Nil _ -> Single xs
  | Cons (h, t) -> Ncons (xs, tails t)

let program xs = sum (map product (tails xs))